source/common/buffer/buffer_impl.cc

```diff
@@ -169,25 +168,36 @@ void OwnedImpl::drain(uint64_t size) {
   }
 }
 
-uint64_t OwnedImpl::getRawSlices(RawSlice* out, uint64_t out_size) const {
+uint64_t OwnedImpl::getAtMostNRawSlices(RawSlice* out, uint64_t out_size) const {
   uint64_t num_slices = 0;
   for (const auto& slice : slices_) {
+    if (num_slices >= out_size) {
+      break;
+    }
+
     if (slice->dataSize() == 0) {
       continue;
     }
     if (num_slices < out_size) {
       out[num_slices].mem_ = slice->data();
       out[num_slices].len_ = slice->dataSize();
     }
-    // Per the definition of getRawSlices in include/envoy/buffer/buffer.h, we need to return
-    // the total number of slices needed to access all the data in the buffer, which can be
-    // larger than out_size. So we keep iterating and counting non-empty slices here, even
-    // if all the caller-supplied slices have been filled.
     num_slices++;
   }
   return num_slices;
 }
 
+void OwnedImpl::getRawSlices(RawSliceVector& raw_slices) const {
+  raw_slices.reserve(slices_.size());
```

Comment:
Would you lose performance by returning the RawSliceVector? Or does RVO guarantee that would be just as efficient.

It would be more readable and "modern" at call-sites IMO if you could write:
```
  Buffer::RawSliceVector slices = buffer.getRawSlices();
```

Reply:
InlineVector is not movable or copyable.

> Would you lose performance by returning the RawSliceVector? Or does RVO guarantee that would be just as efficient.
> 
> It would be more readable and "modern" at call-sites IMO if you could write:
> 
> ```
>   Buffer::RawSliceVector slices = buffer.getRawSlices();
> ```

I missed that InlinedVector actually has a move constructor.  Let me prototype that...

Reply:
If it has a move constructor I would definitely do that. I suspect that modern compilers would probably do it in many cases without any copy at all (even for the inline bytes) but I'm not sure given it's through a virtual interface. One option would be to have a non-virtual wrapper in the header which does the stack allocation and return which I think would almost definitely use RVO.

Reply:
Yeah RVO should guarantee that no bytes are actually moved, but maybe it needs the move constructor to compile. Hopefully that just works.

Reply:
New version out.  Being able to use InlinedVector and call getRawSlices directly in for-range loops is great.

getAtMostNRawSlices and numSlicesComputedSlowly are also gone.